core/src/main/java/uk/gov/justice/services/core/handler/HandlerMethod.java

```diff
@@ -52,9 +56,23 @@ public HandlerMethod(final Object object, final Method method, final Class<?> ex
                     format("Handles method must have exactly one parameter; found %d", parameterTypes.length));
         }
 
-        if (!isEnvelope(parameterTypes[0])) {
+        final Class<?> envelopeType = parameterTypes[0];
+
+        if (!isEnvelope(envelopeType)) {
             throw new IllegalArgumentException(
-                    format("Handler methods must take an JsonEnvelope as the argument, not a %s", parameterTypes[0]));
+                    format("Handler methods must take an JsonEnvelope or Envelope<T> as the argument, not a %s", envelopeType));
+        }
+
+        if(!envelopeType.equals(JsonEnvelope.class)) {
+            final Type[] genericParameterTypes = method.getGenericParameterTypes();
+            final Type[] parameters = ((ParameterizedType)genericParameterTypes[0]).getActualTypeArguments();
+            try {
+                envelopeGenericType = forName(parameters[0].getTypeName());
+            } catch (ClassNotFoundException e) {
+                throw new HandlerCreationException(e);
+            }
+        } else {
+            envelopeGenericType = JsonEnvelope.class;
```

Comment:
This seems wrong; when the envelope is a JsonEnvelope, this is no longer the payload type